README
README for Tangram 2.08
~~~~~~~~~~~~~~~~~~~~~~~
Tangram is an object-relational mapper. It makes objects persist in
relational databases, and provides powerful facilities for retrieving
and filtering them.  Tangram fully supports object-oriented
programming, including polymorphism, multiple inheritance and
collections.  It does so in an orthogonal fashion, that is, it doesn't
require your classes to implement support functions nor inherit from a
utility class.

Tangram builds upon DBI, so you'll need to install DBI and at least
one DBD driver if that is not done already. You will also need
Set::Object, Scalar::Util, and Data::Lazy 0.6.  These modules are
available from CPAN, though currently Data::Lazy 0.6 has not been
indexed, as the namespace hasn't been handed over - this will probably
need to be installed manually.

If you intend to run the regression tests (recommended), you will also
need to prepare an empty database before starting the installation
procedure, and have Test::More installed.

For more information, see the included POD files and
http://tangram.utsl.gen.nz/



Changes
Revision history for Perl extension Tangram.

0.01  Sun Mar 21 10:46:45 1999
	- original version; created by h2xs 1.18

0.03  Stardate: unknown
	- first public alpha release

0.08  Stardate: unknown
	- second public alpha release

0.09  Stardate: unknown

   - split Tangram.pm into several modules
   - replaced @ISA with 'use base'
   - replaced default typetags with automounting
   - prefetch now takes either class or remote
   - implemented prefetch for IntrArray and IntrSet
      *::prefetch still need real-life testing,
      especially when class(coll) <> class(item)

1.00  Nov 07 1999 - first mature release version
	- deploy() now also works directly on DBI handles
	- deploy() allows specification of SQL types via the schema hash
	- intrusive collections allow element-to-container navigation
	- improve speed of read_object
	- fixed various bugs
	- rewrote part of, and expanded documentation

1.01  Nov 12 1999 - interim version
	- added (raw) mapping between Perl strings and date/time SQL types

1.02  Nov 14 1999 - interim version
	- added support for aggregated refs, arrays and sets

1.03  Dec 28 1999 - interim version
	- added support for customizing object & class id types, etc
	- as a consequence Tangram should now work with PostgreSql
	- fixed bug: erase() failed in presence of stateless leaf classes
	- temporarily removed RawDate and RawTime tests because M$ SQL
	  cannot handle them; those modules are still present though

1.04  Jan 03 2000 - interim version
	- fixed two bugs related to explicit tables

1.05  Jan 03 2000 - interim version
	- fixed: Storage::prefetch now works with inherited members
	- fixed: _insert() ignored collections in stateless classes

1.06  Feb 29 2000 - higly experimental interim version
	- added typechecks on collection members
	- started cleaning up autovivification of colls upon save
	- documented like()
	- documented extended select options retrieve, order & filter
	- changed _limit into the Limit in test suite
	- added temporary hack in Makefile.PL to support Access
	- added mapping for flat arrays
	- factorized code common to Array & FlatArray

1.07  Feb 29 2000 - interim version
	- adapted test suite for Sybase and other NULL-picky databases
	- added experimental support for vendor extensions (Dialect)
	- added skeleton Dialect::Sybase with support for datepart
	- did extensive rework on Expr instantiation
	- began documenting Type
	- added Storage::reload()

1.08  Mar 26 2000 - interim version
	- batch allocation of object ids
	- support for Lobs contributed by marian.kelc@ruhr-uni-bochum.de
	- Storage::reset()

1.09  ? - interim version
	- String calls DBI for correct quoting, as Marian Kelc suggested
	- fixed Array problem when removing elements from middle of collection
	- (?)

1.10  Apr 19 2000 - interim version
	- changed reset() into unload() and added support for partial unloading
	- tx_do() can now propagate arguments to closure/anon-sub;
	  insert(), update() and erase() pass anon-subs to tx_do to prevent
	  leaks (thanks Radu Greab)

1.11  May 29 2000 - public version
	- ordering result set by descending order is now supported
	- die() on attempt to load from an invalid id
	- removed silly table id allocation limit (t1000) in Coll.pm

1.13  Jul 11 2000 - interim version
	- new FlatHash and DMDateTime mappings contributed by Gabor Herr
          <herr@iti.informatik.tu-darmstadt.de>
	- some more doc in Type.pod
	- date & time doc moved to DateTime.pod
	- deploy() puts \n between instructions in trace

1.15  Jul 21 2000 - interim version
	- new PerlDump mapping contributed by Gabor Herr
          <herr@iti.informatik.tu-darmstadt.de>

1.16  Jul 25 2000 - interim version
	- fixed deploy() bug: didn't take col names into account for Array, Set and Hash
	- fixed query_expr bug in Scalar and Ref: used field name instead of col name

1.17  Jul 28 2000 - interim version
	- Gabor's patches on deep updates

1.18  Jul 28 2000 - stable version
	- weakrefs!
	- tuples
	- mysql tx acquired lock regardless of tx nesting

1.19  Jul 28 2000 - stable version
	- improved id allocation for mysql
	- bitwise operators for mysql
	- manual class id allocation
	- install now works in the absence of DBI and Set::Object
	- Makefile.PL prerequisites work now
	- Guided Tour is now generated & can be run
	- added support for LIMIT in selects
	- deploy() didn't take explicit oid/cid type into account; fixed
	- deploy() wrongly defaulted to NULLable types; fixed
	- class id can now be specified
	- oid and cid now default to INTEGER
	- fixed error in Expr.pod but I don't remember what it was	
	- minor improvement to Schema.pod
	- added Springfield.pm for users to play with
	- added support for SELECT SUM

2.00  Feb ?? 2001 - first S.O.L. release
	- added copyright notices
	- added a copy of the GPL
	- implemented inlined exports
	- INSERT and UPDATE now make use of prepared statements when available
	- re-executable Cursors
	- changed database layout, Class Table is gone and 'control' table in there
	- object ids are no longer allocated on a class basis
	- introduced Tangram::Relational
	- dialect are now done via derivation
	- stored object id no longer contains type id
	- introduced Relational::Engine
	- flexible mapping

2.05pre1 May 2003 - SamV - first "not the momma" release

	- Patches from lots of people, don't know who 'coz the honourable
	  grand master JLL applied a lot of them, I'll try to summarize
	  here:
	
		- back-porting pieces for 5.00503 (that was me)
		- a more elegant solution for the
		  Tangram::AbstractArray/AbstractHash corner case than
		  the one I submitted :-)
		- code that (possibly) completes Tangram::Hash support (I
		  haven't written a test case to confirm this yet)
		- A fix to Tangram::AbstractSet that again caters to some
		  very strange corner cases (this was me too)
		- Working implementation of the Tangram::DMDateTime type
		  mapping, a good thing that running a persistent

Makefile.PL
use strict;

use ExtUtils::MakeMaker;

my %required =
    (
     'Set::Object' => 1.08,
     "Test::More" => 0,
     #"Date::Manip" => 0,
     "Time::Piece" => 0,
     "Class::Date" => 0,
     #"DateTime" => 0,
     "Scalar::Util" => 0,
     "Data::Lazy" => 0.6,
     DBI => 0,
    );

my %required_soft = map{$_=>1} qw( Scalar::Util Time::Piece Class::Date Data::Lazy );
my %required_hard = map{$_=>1} qw( Set::Object  DBI Set::Scalar );

my ($skip_tests, $bomb_out);
while (my ($module, $min_ver) = each %required) {

    my $mod_ver = "$module".($min_ver ? " $min_ver" : "");
    eval "use $mod_ver;";
    if($@) {
	if (exists $required_hard{$module}) {
	    $bomb_out = 1;
	    print STDERR ("$mod_ver not found.\n");
	} elsif ( exists $required_soft{$module} ) {
	    print STDERR ("$mod_ver not found.  Some tests and/or "
			  ."functionality may be missing.\n");
	} else {
	    $skip_tests = 1;
	    print STDERR ("$mod_ver not found.  You will not be able "
			  ."to run the test suite.\n");
	}
    }
}

do { $skip_tests = 1; goto NOTESTS } if $skip_tests or $bomb_out;

use lib '.';

sub yes
{
    print ' (Y/n) ';
    return <STDIN> =~ /^(Y(e(s)?)?|A(YE|II+!*))?\n?$/i;
}

sub yeah_no  # it's an antipodean thing
{
    print ' (N/y) ';
    return <STDIN> =~ /^(Y(e(s)?)?|A(YE|II+!*))\n?$/i;

}

print q{Do you plan to run the test suite?
(you will need to set up an *EMPTY* database)};

do { $skip_tests = 1; goto NOTESTS } unless yes();

my $configured;

if ($ENV{TANGRAM_CONFIG})
{
   print qq{
You have set TANGRAM_CONFIG to $ENV{TANGRAM_CONFIG}.
Should I use it?};
   $configured = yes();
   unless ($ENV{TANGRAM_CONFIG} eq "t/CONFIG") {
      open TCONF, "$ENV{TANGRAM_CONFIG}"
	  or die "failed to open $ENV{TANGRAM_CONFIG} for reading; $!";
      unlink "t/CONFIG";
      open CONFIG, ">t/CONFIG"
	  or die "failed to open t/CONFIG for writing; $!";
      while (<TCONF>) { print CONFIG }
      close TCONF;
      close CONFIG;
   }
}

if (!$configured && -e 't/CONFIG')
{
   print q{
It looks like there is a 't/CONFIG' file already. It probably contains
connection information from a previous installation. Should I use it?};
   $configured = yes();
}

unless ($configured)
{
   print q{
Please give me the login and password for accessing the test database.
I must be able to create and drop tables in that database.};

   print "\n1) DBI connect string (omit the \'DBI:\' part): ";
   my $cs = <STDIN>;
   chop $cs;

   $cs = "dbi:$cs" unless $cs =~ /^dbi\:/i;

   my ($use_tx, $use_subsel, $table_type) = (1, 1);

   if ($cs =~ m/:mysql:/i) {
       print q{
You have selected the MySQL back-end.  Normally, subselects and
transactions are disabled for this database.  However, if you are
using MySQL-Max, or some other MySQL version with InnoDB support
compiled in and configured, you can enable the transaction support for
the test suite.  If you want to use it in your programs, you'll have
to set no_tx = 0 in the options to Tangram::Storage->connect() (or
hope that the auto-detection of the feature works), and table_type =
InnoDB in the Schema.  See Tangram::Schema and Tangram::Storage for
more information.

Alternatively, if you are running MySQL 4.1 or later, you can enable
sub-select tests.  It is not possible to use sub-selects with InnoDB
table types.

Use InnoDB tables};
       unless (yeah_no()) {
	   $use_tx = 0;

	   print "Use sub-selects";
	   unless (yeah_no()) {
	       $use_subsel = 0;
	   }
       } else {
	   $table_type = "InnoDB";
       }
   }

   my ($user, $passwd);
   if ( $cs =~ /:sqlite:/i ) {
       print q{
You have selected the SQLite back-end.  Sub-selects will be disabled.
};
       $use_subsel = 0;
       $user = $passwd = "";
   } else {

       print "2) Login: ";
       $user = <STDIN>;
       chop $user;

       print "3) Password: ";
       $passwd = <STDIN>;
       chop $passwd;

   }
   
   print <<'MSG';

Thank you. I am going to save this information to 't/CONFIG'.
If you have given me sensitive information, make sure to destroy
the file when the tests have been completed.
MSG

   open CONFIG, '>t/CONFIG' or die "Cannot create 't/CONFIG', reason: $!";
   print CONFIG "$cs\n$user\n$passwd\ntx_do = $use_tx\nsubselects = $use_subsel\n".($table_type?"table_type = $table_type\n":"");
   close CONFIG;
}

NOTESTS:
my $tests;
if ( $skip_tests ) {
    print "TEST SUITE IS DISABLED\n";
    $tests = "t/no_tests.t";
} else {
    eval "use Test::Manifest";


    if ( $@ ) {
	open TESTS, "<t/test_manifest" or die $!;
	$tests = join " ", map { chomp; "t/$_" } <TESTS>;
	close TESTS;
    }

    else {
	* ExtUtils::MM_Any::test_via_harness = sub
	    {
		my($self, $perl, $tests) = @_;

		return (qq|\t$perl "-MTest::Manifest" | .
			qq|"-e" "run_t_manifest(\$(TEST_VERBOSE), |.
			qq|'\$(INST_LIB)', | .
			qq|'\$(INST_ARCHLIB)')"\n|);
	    };
    }
}

WriteMakefile(
	      'NAME'	=> 'Tangram',
	      'VERSION_FROM' => 'lib/Tangram.pm', # finds $VERSION
	      PREREQ_PM => \%required,
	      test => { TESTS => $tests },
);
